### PR TITLE
@planetarium/tx for creating transactions in JavaScript

### PR DESCRIPTION
**Pull request by dahlia**
*Addresses #2258.*

This adds the following components:

 -  *@planetarium/tx*, a new npm package for creating Libplanet transactions in JavaScript/TypeScript.  (However, it supports only unsigned transactions with system actions yet.)
 -  `planet tx analyze`, a new command to dump a (Bencodex-encoded) transaction into a human-readable JSON tree.  It also can take unsigned transactions as well.  It's used for integration testing of `encodeUnsignedTxWithSystemAction()`.

It's ready for review now, except that there are still few CI things to work on.  I'm going to add few more commits for that work.

### CODE REVIEW

**Comment by longfin** [comment]
Considering [the use case of chorno]((https://github.com/planetarium/pn-chrono/blob/997a38335a89e3d19ebb3cac1298fea76c12a0ce/background/src/api/graphql.js#L107)), It would be nice if we have ts version of [`FungibleAssetValue.Parse()`](https://docs.libplanet.io/0.41.1/api/Libplanet.Assets.FungibleAssetValue.html#Libplanet_Assets_FungibleAssetValue_Parse_Libplanet_Assets_Currency_System_String_).

**Comment by longfin** [comment]
When we published this library with tx creation features, it would be better that deprecate `unsignedTransaction`, `bindSignature` queries. (I realized that they couldn't parse system actions due to `T` 😢 ). cc @moreal

File: @planetarium/tx/test/tx/unsigned.test.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
+import { join } from "node:path";
+import { encode } from "bencodex";
+import { execa } from "execa";
+import { expect, test } from "vitest";
+import { encodeUnsignedTxWithSystemAction} from "../../src/tx/unsigned";
+import { key1, key2 ,FOO} from "./fixtures";
```
Comment (tkiapril) [inline @planetarium/tx/test/tx/unsigned.test.ts:6]:
```suggestion
import { key1, key2 ,FOO } from "./fixtures";
```

**Review by BasixKOR — COMMENTED** [review]
Some notes:
- As JavaScript users won't use the tsc in most cases, it may be a good idea to add some foolproof checks instead of fully relying on TypeScript.
  - I usually do this by calling `typeof` in place of null checks.
- You may want to look into TypeDoc for generating the documentation.

File: @planetarium/tx/src/key.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { Encodable } from "bencodex";
+
+export type PublicKey = Uint8Array;  // TODO: proper type definition
```
[the remark this answers is not in the document]
Comment (dahlia) [reply @planetarium/tx/src/key.ts:3]: If branded types mean nominal types `PublicKey` is not a distinct nominal type, but just an alias of `Uint8Array` — TypeScript allows we pass an `Uint8Array` instance to a parameter with `PublicKey`.  That's why I leave a `TODO` comment at present.

File: @planetarium/tx/src/key.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+export function encodePublicKey(publicKey: PublicKey): Encodable {
+  if (publicKey.length < 1) {
+    throw new TypeError("Public key must not be empty.");
+  }
+  return publicKey.buffer;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply @planetarium/tx/src/key.ts:9]: That's one thing that has repeatedly confused me.  Anyway, the type check allows me to assign an `ArrayBuffer` to `Encodable`, which seemingly contains `ArrayBuffer` besides `Uint8Array | Buffer` probably?

File: @planetarium/tx/src/tx/metadata.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+const PUBLIC_KEY_KEY = Buffer.from([0x70]); // 'p'
+const TIMESTAMP_KEY = Buffer.from([0x74]); // 't'
+
+export function encodeTxMetadata(
+  metadata: TxMetadata
+): Map<string | Buffer,Encodable> {
+  const updatedAddresses = [...metadata.updatedAddresses].map(encodeAddress);
+  const timestamp = metadata.timestamp.toISOString()
+    .replace(/Z$/, "000Z");
+  const dict = new Map<string | Buffer, Encodable>([
```
[the remark this answers is not in the document]
Comment (dahlia) [reply @planetarium/tx/src/tx/metadata.ts:32]: Currently no reason, but if Libplanet decides to use Unicode string keys for tx payloads in the future this function's return type also should be able to include them too.

File: @planetarium/tx/src/key.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (BasixKOR) [reply @planetarium/tx/src/key.ts:9]: [Yep](https://github.com/disjukr/bencodex/blob/2116784deab313e82b1819ee48e60f39d08278cf/src/index.ts#L5). It seems like they are converted to Buffer by the serialization library, so I don't think there are big concerns.

File: @planetarium/tx/src/action.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      type_id: 0,
+      values: encodeMint(action),
+    };
+  case "transfer":
+    return {
+      type_id: 1,
+      values: encodeTransfer(action),
+    };
+  default:
+    throw new TypeError(`Unknown action type: ${action["type"]}`);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply @planetarium/tx/src/action.ts:20]: Copilot has a poor sense of style.  😜

[the remark this answers is not in the document]
Comment (dahlia) [reply @planetarium/tx/src/action.ts:20]: I tried to apply your suggestion, but TypeScript gave me an error T2339[^1], as this `switch` covers every case of union type `SystemAction`.

[^1]: “Property 'type' does not exist on type 'never'.”

[the remark this answers is not in the document]
Comment (BasixKOR) [reply @planetarium/tx/src/action.ts:20]: Oops, I see.

File: @planetarium/tx/src/assets.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+export function encodeCurrency(currency: Currency): Encodable {
+  const minters: Encodable[] | null
+    = currency.minters === null
+      ? null
+      : [...currency.minters].sort(compareUint8Array).map(encodeAddress);
+  const serialized: Encodable = {
+    ticker: currency.ticker,
+    decimals: currency.decimalPlaces,
+    minters,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply @planetarium/tx/src/assets.ts:24]: It's a part of the specification. Field `totalSupplyTrackable` has to either have `true` or be not there; `totalSupplyTrackable: false` is disallowed due to backward compatibility.  (Note that `totalSupplyTrackable` was introduced recently.)

[the remark this answers is not in the document]
Comment (dahlia) [reply @planetarium/tx/src/assets.ts:24]: The set of keys can vary according to data.  Such variance is due to Libplanet's backward compatibility…

File: @planetarium/tx/src/binary.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+
+export function compareUint8Array(a: Uint8Array, b: Uint8Array) {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply @planetarium/tx/src/binary.ts:2]: It's used for normalizing a set of binary data (e.g., addresses) in Bencodex.  As Bencodex does not have first-class sets, we need to encode sets into sorted arrays.

**Comment by dahlia** [comment]
@BasixKOR I further configured eslint settings and added a pre-commit hook for *@planetarium/tx*.  Could you take a look on b0a3f768911de88cf2e63fc762722f5c032e0635?

File: @planetarium/tx/src/binary.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+  const length = Math.min(a.length, b.length);
+  for (let i = 0; i < length; i++) {
+    if (a[i] !== b[i]) return a[i] - b[i];
```
Comment (tkiapril) [inline @planetarium/tx/src/binary.ts:5]: Just a curiosity, but is the reason for returning the difference of two bytes about interpreter optimization?

Comment (dahlia) [reply @planetarium/tx/src/binary.ts:5]: Actually there's no reason behind this.  It's just my habit.

File: Libplanet.Extensions.Cocona/Commands/TxCommand.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            bool unsigned = false,
+            [Option(
+                'i',
+                Description = "Terminate the program with a zero exit code (success) even if " +
+                    "the signature is invalid.  However, it still prints the error message to " +
+                    "the standard error, and the exit code is still non-zero if the transaction " +
+                    "is invalid in other ways.  This option is mutually exclusive with " +
+                    "-u/--unsigned."
+            )]
+            bool ignoreSignature = false
```
[the remark this answers is not in the document]
Comment (dahlia) [reply Libplanet.Extensions.Cocona/Commands/TxCommand.cs:36]: Amended: 9463713b55e336c7bbdb94674ce5d7c038889a53.

File: Libplanet.Extensions.Cocona/Commands/TxCommand.cs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            {
+                sourceName = $"file {file}";
+                try
+                {
+                    bytes = File.ReadAllBytes(file);
+                }
+                catch (Exception)
+                {
+                    throw new CommandExitedException(
+                        $"The file {file} does not exist.",
```
[the remark this answers is not in the document]
Comment (dahlia) [reply Libplanet.Extensions.Cocona/Commands/TxCommand.cs]: Amended: ced81a8fd199318a8ef22c6c49e9056dc6757410.

File: .github/bin/dist-npm.sh [hunk 11]
```diff
@@ -18,14 +18,9 @@ else
...
-         "publishing it to npm." > /dev/stderr
-    continue
-  fi
-  pushd "./$project/"
+for npmpkg in "${npm_packages[@]}"; do
+  for tgz in "./$npmpkg"/*.tgz; do
     # shellcheck disable=SC2086
-    npm publish --access=public $dry_run
-  popd
+    npm publish --access=public $dry_run "$tgz"
```
Comment (BasixKOR) [inline .github/bin/dist-npm.sh:24]: I *think* I would skip `pack` by calling `publish` without the tarball argument but not a dealbreaker.

File: Libplanet.Extensions.Cocona/Commands/TxCommand.cs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+                    );
+                }
+
+                if (!invalid)
+                {
+                    Console.Error.WriteLine(
+                        "The signature from the transaction {0} is valid.",
+                        sourceName
+                    );
+                    invalid = !ignoreSignature;
```
Comment (tkiapril) [inline Libplanet.Extensions.Cocona/Commands/TxCommand.cs:126]: Wouldn't this line make the command return -1 when the signature is valid and ignoreSignature is false?

File: @planetarium/tx/test/address.test.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      (addresses: Uint8Array[]) => {
+        const set = new Set(addresses);
+        const checks = new Set(addresses.map(hex));
+        const encoded = encodeAddressSet(set);
+        if (!(encoded instanceof Array)) return false;
+        let prev: ArrayBuffer | null = null;
+        for (const v of encoded) {
+          if (!(v instanceof ArrayBuffer)) return false;
+          if (v.byteLength != 20 || !checks.has(hex(v))) return false;
+          if (prev != null && compareUint8Array(prev, v) >= 0) return false;
```
Comment (BasixKOR) [inline @planetarium/tx/test/address.test.ts:63]: This test feels like a reimplementation of `encodeAddressSet`, although I'm not so sure what would be a better fit here.

How about testing the order to be consistent for every randomly-ordered set?

**Review by BasixKOR — APPROVED** [review]
There are a few codes I'd like to revisit, but none of them is a dealbreaker.

**Review by tkiapril — COMMENTED** [review]
Just a thought, but I wonder if the name of encode~~~() functions being similar to the bencodex encode() function, as the former returns a marshalled json dictionary whereas the latter returns a bencodex encoded string, especially since when `@planetarium/tx` is being used I think most of the applications would be using bencodex as well in the same context.

File: .github/bin/dist-npm.sh (hunk 11 above)
Comment (dahlia) [reply .github/bin/dist-npm.sh:24]: That's what I intended.  I want to build tarballs even if they shouldn't be published to npm.  It's useful for debugging CI/CD.

File: @planetarium/tx/test/address.test.ts (hunk 13 above)
Comment (dahlia) [reply @planetarium/tx/test/address.test.ts:63]: > This test feels like a reimplementation of `encodeAddressSet`,

Whereas `encodeAddressSet()` builds an ordered array, it does not build data.  Instead, it only checks if every element is transitive and unique in the array.
 
> How about testing the order to be consistent for every randomly-ordered set?

What do you mean by random ordering?  If you mean shuffle, how an input of `new Set()` is ordered does not matter as `Set` in itself does not have ordering.  Instead, we need to ensure if the result of `encodeAddressSet()` is properly ordered.

Comment (BasixKOR) [reply @planetarium/tx/test/address.test.ts:63]: Oops I must've misread.

File: CHANGES.md [hunk 14]
```diff
@@ -20,9 +25,14 @@ To be released.
...
- -  `RoutingTable` became to implement `IRoutingTable` interface.
-    [[#2046], [#2229]]
+ -  System actions became to have methods to check equality.  [[#2294]]
+     -  `Mint` became to implement `IEquatable<Mint>`.
+     -  `Mint` became to implement `IEquatable<IAction>`.
+     -  `Transfer` became to implement `IEquatable<Transfer>`.
+     -  `Transfer` became to implement `IEquatable<IAction>`.
+ -  (Libplanet.Net) Added `IRoutingTable` interface.  [[#2046], [#2229]]
+ -  (Libplanet.Net) `RoutingTable` became to implement `IRoutingTable`
+    interface.  [[#2046], [#2229]]
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md]: Adjusted: 27e1ab577f8c281077e5706743e3118abb13efb1.

File: CHANGES.md [hunk 15]
```diff
@@ -31,36 +41,61 @@ To be released.
...
 [#2046]: https://github.com/planetarium/libplanet/issues/2046
 [#2131]: https://github.com/planetarium/libplanet/issues/2131
 [#2229]: https://github.com/planetarium/libplanet/pull/2229
 [#2236]: https://github.com/planetarium/libplanet/pull/2236
+[#2294]: https://github.com/planetarium/libplanet/pull/2294
 [#2322]: https://github.com/planetarium/libplanet/pull/2322
 [`System.Text.Json.JsonSerializer`]: https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer
 [custom converters]: https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to
+[@planetarium/tx]: https://www.npmjs.com/package/@planetarium/tx
+[Bencodex JSON Representation]: https://github.com/planetarium/bencodex/blob/4a92a98e859a54bc6e9617d4dd3035250fe69a86/CHANGES.md
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md]: Amended: 141c31aebe0bd660e69128b66cea881be4cab770.

File: Libplanet.Extensions.Cocona/Commands/TxCommand.cs [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+        public int Analyze(
+            [Argument(
+                Description = "The file path of the transaction to analyze.  If " +
+                    "a hyphen (-) is given it reads from the standard input (if you want to read " +
+                    "just a file named \"-\", use \"./-\" instead)."
+            )]
+            string file,
+            [Option(
+                'u',
+                Description = "The transaction to analyze is unsigned.  This option is " +
```
[the remark this answers is not in the document]
Comment (dahlia) [reply Libplanet.Extensions.Cocona/Commands/TxCommand.cs]: Amended: 34bb8b4e39425f74ef0b4cc2fdda99dae23073e7.

File: Libplanet.Extensions.Cocona/Commands/TxCommand.cs (hunk 12 above)
Comment (dahlia) [reply Libplanet.Extensions.Cocona/Commands/TxCommand.cs]: Amended: 34bb8b4e39425f74ef0b4cc2fdda99dae23073e7.

**Review by tkiapril — APPROVED** [review]

**Pull request merged**